management/commands/aggregate_subject_trials.py

```diff
@@ -0,0 +1,526 @@
+            else:
+                old_trials = pd.read_parquet(old_aggregate)
+                rerun |= set(sessions['id'].astype(str).unique()) != set(old_trials['session'].unique())
+                # If there are more sessions to extract now, do the re-run
+                to_extract = {str(eid) for eid, _, note in outcomes if note == 'INITIALIZED'}
+                rerun |= to_extract > set(old_trials['session'].unique())
+                if rerun:
+                    logger.info('Set of sessions changed')
+                else:
+                    aggregate_hash = self.make_aggregate_hash(to_extract, input_files=input_files)
+                    rerun |= aggregate_hash != (qs.first().json or {}).get('aggregate_hash')
+                    if rerun:
+                        logger.info('Aggregate hash changed')
+
+        if not rerun:
+            logger.info('Aggregate hash unchanged; exiting')
+            return
+
+        # Generate aggregate trials table
+        all_trials, outcomes = generate_trials_aggregate(all_tasks, outcomes)
+        outcomes = pd.DataFrame(outcomes, columns=['session', 'task number', 'notes'])
+        outcomes.drop_duplicates(['session', 'task number'], keep='last', inplace=True)
+        # Add start_times column to trials table
+        start_times = sessions.astype({'id': str}).set_index('id')['start_time'].rename('session_start_time')
+        all_trials = all_trials.merge(start_times, left_on='session', right_index=True, sort=False)
+        assert set(all_trials.columns) == set(EXPECTED_KEYS), 'unexpected columns in aggregate trials table'
+
+        if out_file.exists():
+            logger.warning(('(DRY) ' if dry else '') + 'Output file already exists, overwriting %s', out_file)
+        if dry:
+            out_file = out_file.with_name(f'{out_file.stem}.DRYRUN{out_file.suffix}')
+        # Save to disk
+        out_file.parent.mkdir(parents=True, exist_ok=True)
+        all_trials.to_parquet(out_file)
+        assert out_file.exists(), f'Failed to save to {out_file}'
+        assert (file_size := out_file.stat().st_size) > 0
+        assert not pd.read_parquet(out_file).empty, f'Failed to read-after-write {out_file}'
+        md5_hash = hashfile.md5(out_file)
+        # Save outcome log
+        outcomes.to_csv(out_file.with_name(f'_ibl_subjectTrials.log{".DRYRUN" if dry else ""}.csv'))
+
+        # Create aggregate hash
+        successful_sessions = outcomes['session'][outcomes['notes'] == 'SUCCESS'].unique().tolist()
+        aggregate_hash = self.make_aggregate_hash(successful_sessions, input_files=input_files)
+
+        if dry:
+            logger.info('Dry run complete: %s aggregate hash = %s', out_file, aggregate_hash)
+            return
+
+        # Create dataset
+        self.register_dataset(out_file, file_hash=md5_hash, file_size=file_size, aggregate_hash=aggregate_hash, user=user)
+
+    @staticmethod
+    def files2datasets(file_list):
```

Comment:
This method appears to be unused

Reply:
Yes I'm embarrassed to admit that I wrote that method before realizing I could just get the dataset from the UUID in the filename. I'll move it somewhere else.